csrc/dynamic_transform.cpp

```diff
@@ -385,6 +689,224 @@ void DynamicTransformConcretizer::concretize() {
+  //
+  //    T0          T2
+  //     |           |
+  //   PadOp       PadOp
+  //       \       /
+  //         CatOp
+  //           |
+  //          T3
+  //
+  // This is done by simply calling the cat() command with only {T0, T2}.
+  auto pad_uses = pop->out()->uses();
+  if (pad_uses.size() == 1 && pad_uses[0]->isA<CatOp>()) {
+    auto cop = pad_uses[0]->as<CatOp>();
+    std::vector<TensorView*> nonempty_inputs;
+    for (auto inp : cop->inputs()) {
+      // Each "input" to CatOp is a pad() of the corresponding _actual_
+      // input. Here we peel off the pad op to collect the non-padded cat
+      // inputs.
+      auto padded_inp_tv = inp->as<TensorView>();
+      TORCH_INTERNAL_ASSERT(
+          padded_inp_tv->definition() &&
+              padded_inp_tv->definition()->isA<PadOp>(),
+          "Input to cat should have definition that is a PadOp");
+      auto inp_pad_op = padded_inp_tv->definition()->as<PadOp>();
+      auto inp_tv = inp_pad_op->in()->as<TensorView>();
+
+      if (inp_pad_op != pop) {
+        // We could remove other empty tensors here while we're at it. They will
+        // get removed by further passes anyway though as tv ranges over all
+        // empty tensors.
+        nonempty_inputs.push_back(inp_tv);
+      }
+    }
+    auto old_cat = cop->output(0)->as<TensorView>();
+    auto new_cat = nonempty_inputs.size() == 1
+        ? set(nonempty_inputs[0])
+        : cat(nonempty_inputs, cop->concatenatedDim());
+    replaceTV(old_cat, new_cat);
+  } else { // Replace pads that are not part of CatOps with full()
+    auto out_shape = tensor_shape(out);
+    // Wherever there is a zero in the input, we will replace the original
+    // output extent so that we no longer reference the now-zero input
+    // extent
+    for (auto ax : empty_axes) {
+      auto pad_widths = pop->getPadWidths((int)ax);
+      out_shape[ax] = add(pad_widths.first, pad_widths.second);
+    }
+    replaceWithFull(out, out_shape, pop->value());
+  }
+}
+
+void DynamicTransformConcretizer::removeEmptyBranches() {
+  auto fusion = FusionGuard::getCurFusion();
+  for (const auto& empty_tv_descr : info_->getEmptyTensors()) {
+    auto tv = info_->initialInfo()->lookUpTV(empty_tv_descr.tv_name);
+    auto rfactor = TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+    std::vector<Val*> new_shape;
+    new_shape.reserve(rfactor.size());
+    for (auto id : rfactor) {
+      new_shape.push_back(id->getMaybeExpandedExtent());
+    }
+    for (auto ax : empty_tv_descr.empty_axes) {
+      // Hard-code zero extent for empty axes. This lets us detect empty input
+      // and output tensors during scheduling/execution.
+      registerConcretization(new_shape[ax], fusion->zeroVal());
+      new_shape[ax] = fusion->zeroVal();
+    }
+
+    for (auto use : tv->uses()) {
+      replaceEmptyUse(use, empty_tv_descr.empty_axes);
+    }
+    if (!tv->isFusionInput()) {
+      replaceWithFull(tv, new_shape);
```

Comment:
If all uses of `tv` is replaced by `replaceEmptyUse`, is this `replaceWithFull` necessary? I think not, but it would just become a dead full op, right?